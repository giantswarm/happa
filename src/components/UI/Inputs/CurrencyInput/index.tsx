import PropTypes from 'prop-types';
import * as React from 'react';
import styled from 'styled-components';

import { InputElement } from '../Input';

const InputWrapper = styled.div<{ disabled?: boolean }>`
  border: 1px solid
    ${({ theme, disabled }) =>
      disabled ? theme.colors.darkBlueLighter2 : theme.colors.darkBlueLighter1};
  border-radius: ${({ theme }) => theme.border_radius};
  color: ${({ theme, disabled }) =>
    disabled ? theme.colors.gray : theme.colors.whiteInput};
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledInput = styled(InputElement)`
  border-radius: ${({ theme }) => theme.border_radius};
  border: 0;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
`;

const CurrencyLabel = styled.div<{ disabled?: boolean }>`
  color: ${(props) => props.theme.colors.whiteInput};
  background: ${({ theme, disabled }) =>
    disabled ? theme.colors.darkBlueLighter2 : theme.colors.darkBlueLighter1};
  color: ${({ theme, disabled }) =>
    disabled ? theme.colors.white3 : theme.colors.white5};
  padding: 8px 10px;
  line-height: 1;
`;

const LabelText = styled.label`
  font-size: 14px;
  color: ${(props) => props.theme.colors.white2};
`;

const ValidationError = styled.span`
  font-size: 12px;
  color: ${(props) => props.theme.colors.yellow1};
`;

interface ICurrencyInputProps
  extends Omit<
    React.ComponentPropsWithRef<'input'>,
    'type' | 'onChange' | 'step'
  > {
  currencyLabel?: string;
  precision?: number;
  value?: number;
  label?: string;
  min?: number;
  max?: number;
  labelTextProps?: React.ComponentPropsWithRef<'label'>;
  inputWrapperProps?: React.ComponentPropsWithRef<'div'>;
  rootProps?: Omit<React.ComponentPropsWithRef<'div'>, 'htmlFor'>;
  onChange?: (newValue: number) => void;
  validationError?: string;
}

const CurrencyInput: React.FC<ICurrencyInputProps> = ({
  currencyLabel,
  precision,
  label,
  labelTextProps,
  inputWrapperProps,
  rootProps,
  value,
  id,
  min,
  max,
  onChange,
  validationError,
  disabled,
  ...rest
}) => {
  // eslint-disable-next-line no-magic-numbers
  const step = 1 / 10 ** precision!;

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let valueConvertedToPrecision = formatNumberWithPrecision(
      e.target.valueAsNumber,
      precision!
    );

    if (min) {
      valueConvertedToPrecision = Math.max(valueConvertedToPrecision, min);
    }
    if (max) {
      valueConvertedToPrecision = Math.min(valueConvertedToPrecision, max);
    }

    onChange?.(valueConvertedToPrecision);
  };

  const visibleValue = formatNumberWithPrecision(value ?? 0, precision!);

  return (
    <div {...rootProps}>
      {label && (
        <LabelText {...labelTextProps} htmlFor={id}>
          {label}
        </LabelText>
      )}
      <InputWrapper {...inputWrapperProps} disabled={disabled}>
        <CurrencyLabel disabled={disabled}>{currencyLabel}</CurrencyLabel>
        <StyledInput
          {...rest}
          id={id}
          value={visibleValue}
          type='number'
          step={step}
          onChange={handleChange}
          min={min}
          max={max}
          disabled={disabled}
        />
      </InputWrapper>
      {validationError && (
        <ValidationError>
          <i className='fa fa-warning' /> {validationError}
        </ValidationError>
      )}
    </div>
  );
};

CurrencyInput.propTypes = {
  /**
   * Define a currency to be displayed in the
   * left side of the input.
   */
  currencyLabel: PropTypes.string,
  /**
   * The number of decimal places allowed.
   */
  precision: PropTypes.number,
  /**
   * The controlled value of the input.
   */
  value: PropTypes.number,
  /**
   * The unique identifier of the input.
   */
  id: PropTypes.string,
  /**
   * The label displayed above the input.
   */
  label: PropTypes.string,
  /**
   * The minimum allowed value.
   */
  min: PropTypes.number,
  /**
   * The maximum allowed value
   */
  max: PropTypes.number,
  /**
   * Props to be passed to the label element.
   */
  labelTextProps: PropTypes.object,
  /**
   * Props to be passed to the input element's parent element.
   */
  inputWrapperProps: PropTypes.object,
  /**
   * Props to be passed to the root element.
   */
  rootProps: PropTypes.object,
  /**
   * The value change event handler.
   */
  onChange: PropTypes.func,
  /**
   * An error message to be displayed near the input.
   */
  validationError: PropTypes.string,
  /**
   * Whether the input can be modified or not.
   */
  disabled: PropTypes.bool,
};

CurrencyInput.defaultProps = {
  currencyLabel: '$',
  precision: 0,
};

export default CurrencyInput;

function formatNumberWithPrecision(num: number, precision: number): number {
  if (Number.isNaN(num)) {
    return 0;
  }

  // eslint-disable-next-line no-magic-numbers
  const precisionFactor = 10 ** precision;

  return Math.round(num * precisionFactor) / precisionFactor;
}
