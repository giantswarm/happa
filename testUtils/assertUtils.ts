import { MatcherFunction, Nullish } from '@testing-library/react';

type Query = (matcher: MatcherFunction) => HTMLElement;

/**
 * Decorate an existing query to match text across multiple
 * elements.
 * @param query
 */
export function withMarkup(query: Query) {
  return (text: string) => {
    return query((_, node: Nullish<Element>) => {
      if (!node) return false;

      const hasText = (n: Element) => n.textContent === text;
      const hasTextInChildren = Array.from(node.children).some((child) =>
        hasText(child)
      );

      return hasText(node) && !hasTextInChildren;
    });
  };
}
