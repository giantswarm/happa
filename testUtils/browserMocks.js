// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('isomorphic-fetch');

const localStorageMock = (function () {
  let store = {};

  return {
    getItem: function (key) {
      return store[key] || null;
    },
    setItem: function (key, value) {
      store[key] = value;
    },
    removeItem: function (key) {
      delete store[key];
    },
    clear: function () {
      store = {};
    },

    // replaceWith is not part of the localStorage spec, but useful for us
    // in testing to set the full localStorage to a known set of values.
    replaceWith: function (newStore) {
      store = newStore;
    },
  };
})();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

// Adjust the console logging behaviour during test runs.
/* eslint-disable no-console */
global.console = {
  log: console.log,
  error: console.error,
  warn: jest.fn(), // warnings are surpressed.
  info: console.info,
  debug: console.debug,
};
/* eslint-enable no-console */

global.fetch = fetch;
