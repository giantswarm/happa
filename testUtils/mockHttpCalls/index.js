/***** Constants *****/
export * from './constantsAndHelpers';

/***** Responses *****/

// Info
export * from './info';

// User
export * from './user';

// Auth
export * from './auth';

// Clusters
export * from './clusters';

// Organizations
export * from './organizations';

// Status
export * from './status';

// Apps
export * from './apps';

// Node Pools
export * from './nodePools';

// Releases
export * from './releases';

// App Catalogs
export * from './appCatalogs';
