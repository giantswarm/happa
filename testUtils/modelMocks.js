import * as authSelectors from 'selectors/authSelectors';

jest.mock('model/services/giantSwarm');
jest.mock('selectors/authSelectors');

// eslint-disable-next-line no-import-assign
authSelectors.selectAuthToken = jest.fn(
  (_, state) => state.app.loggedInUser.token
);
