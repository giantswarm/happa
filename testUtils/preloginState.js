import { Providers } from 'shared/constants';
import { MainRoutes } from 'shared/constants/routes';

// This is what the state looks like when someone brand new arrives at the site
// and they are not logged in at all.

export default {
  router: {
    location: {
      pathname: MainRoutes.Login,
      search: '',
      hash: '',
      key: '0fgz24',
    },
    action: 'POP',
  },
  main: {
    selectedOrganization: null,
    selectedClusterID: null,
    firstLoadComplete: false,
    loggedInUser: null,
    permissions: {},
    info: {
      general: {
        installation_name: '',
        availability_zones: null,
        provider: Providers.AWS,
      },
      stats: {
        cluster_creation_duration: {
          median: 0,
          p25: 0,
          p75: 0,
        },
      },
      workers: {
        count_per_cluster: {
          max: null,
          default: 0,
        },
      },
    },
  },
  errors: {},
  metadata: {
    version: {
      current: '',
      new: null,
      lastCheck: 0,
      timer: 0,
    },
  },
  entities: {
    catalogs: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
    },
    clusters: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      v5Clusters: [],
      idsAwaitingUpgrade: {},
    },
    organizations: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      credentials: {
        lastUpdated: 0,
        isFetching: false,
        items: [],
        showForm: false,
      },
    },
    releases: {
      error: null,
      isFetching: false,
      items: {},
    },
    users: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      invitations: {
        lastUpdated: 0,
        isFetching: false,
        items: {},
      },
    },
    nodePools: {
      items: {},
      isFetching: false,
    },
    clusterLabels: {
      requestInProgress: false,
      error: null,
    },
  },
  modal: {
    visible: false,
    templateValues: {},
    template: '',
  },
};
