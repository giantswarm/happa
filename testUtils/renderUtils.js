import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
