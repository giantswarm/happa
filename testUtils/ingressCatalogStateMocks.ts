import { Constants } from 'shared';
import { appCatalogsResponse } from 'testUtils/mockHttpCalls';

const ingressCatalog = appCatalogsResponse.find(
  (c) => c.metadata.name === Constants.INSTALL_INGRESS_TAB_APP_CATALOG_NAME
);
export const catalogsState = {
  entities: {
    catalogs: {
      items: {
        [Constants.INSTALL_INGRESS_TAB_APP_CATALOG_NAME]: ingressCatalog,
      },
    },
  },
};
