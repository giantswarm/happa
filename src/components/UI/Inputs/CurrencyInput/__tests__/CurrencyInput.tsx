import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from 'testUtils/renderUtils';

import CurrencyInput from '..';

describe('CurrencyInput', () => {
  it('renders without crashing', () => {
    renderWithTheme(CurrencyInput, {});
  });

  it('only allows base 10 number values', () => {
    const onChangeMockFn = jest.fn();
    renderWithTheme(CurrencyInput, {
      onChange: onChangeMockFn,
      id: 'test-input',
      label: 'Test input',
    });

    const input = screen.getByLabelText('Test input') as HTMLInputElement;

    fireEvent.change(input, {
      target: {
        value: '35',
      },
    });
    // eslint-disable-next-line no-magic-numbers
    expect(onChangeMockFn).toBeCalledWith(35);

    fireEvent.change(input, {
      target: {
        value: 'asd3f',
      },
    });
    // eslint-disable-next-line no-magic-numbers
    expect(onChangeMockFn).toBeCalledWith(0);

    fireEvent.change(input, {
      target: {
        value: '0xff35dd',
      },
    });
    // eslint-disable-next-line no-magic-numbers
    expect(onChangeMockFn).toBeCalledWith(0);
  });

  it('allows setting the number precision, for handling float values', () => {
    const onChangeMockFn = jest.fn();
    renderWithTheme(CurrencyInput, {
      onChange: onChangeMockFn,
      id: 'test-input',
      label: 'Test input',
      precision: 5,
      value: 13.12312031,
    });

    const input = screen.getByLabelText('Test input') as HTMLInputElement;

    expect(input.value).toBe('13.12312');
    // eslint-disable-next-line no-magic-numbers
    expect(input.valueAsNumber).toBe(13.12312);

    fireEvent.change(input, {
      target: {
        value: '25.1679850404500004',
      },
    });
    // eslint-disable-next-line no-magic-numbers
    expect(onChangeMockFn).toBeCalledWith(25.16798);
  });

  it('respects value constraints', () => {
    const onChangeMockFn = jest.fn();
    renderWithTheme(CurrencyInput, {
      onChange: onChangeMockFn,
      id: 'test-input',
      label: 'Test input',
      min: 1,
      max: 10,
    });

    const input = screen.getByLabelText('Test input') as HTMLInputElement;

    fireEvent.change(input, {
      target: {
        value: '-10',
      },
    });
    // eslint-disable-next-line no-magic-numbers
    expect(onChangeMockFn).toBeCalledWith(1);

    fireEvent.change(input, {
      target: {
        value: '20',
      },
    });
    // eslint-disable-next-line no-magic-numbers
    expect(onChangeMockFn).toBeCalledWith(10);

    fireEvent.change(input, {
      target: {
        value: '1',
      },
    });
    expect(onChangeMockFn).toBeCalledWith(1);
  });

  it('allows displaying a validation error', () => {
    const onChangeMockFn = jest.fn();
    renderWithTheme(CurrencyInput, {
      onChange: onChangeMockFn,
      id: 'test-input',
      label: 'Test input',
      value: 1,
      validationError: 'Naah. Wrong value',
    });

    expect(screen.getByText('Naah. Wrong value')).toBeInTheDocument();
  });
});
