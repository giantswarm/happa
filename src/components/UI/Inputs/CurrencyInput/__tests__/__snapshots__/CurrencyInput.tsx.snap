// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput renders without crashing 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 fWDTlB FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kZzThM"
        >
          <div
            class="StyledTextInput__StyledIcon-sc-1x30a0s-3 fwyRrU"
          >
            <span
              class="StyledText-sc-1sadyjn-0 gxPAiZ"
            >
              $
            </span>
          </div>
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 fnpMSJ"
            step="1"
            type="number"
            value="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
